File: src/Wc3FrameExtLua/Fdf/Highlight.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')

---@type FdfDataClass
local FdfData = require('FdfData') or error('')

--=======
-- Class
--=======

local FdfHighlight = Class.new('FdfHighlight', FdfData)
---@class FdfHighlight : FdfData
local public = FdfHighlight.public
---@class FdfHighlightClass : FdfDataClass
local static = FdfHighlight.static
---@type FdfHighlightClass
local override = FdfHighlight.override
local private = {}

--=========
-- Static
--=========

---@param name string
---@param child FdfHighlight | nil
---@return FdfHighlight
function override.new(name, child)
    isTypeErr(name, 'string', 'name')
    if child then isTypeErr(child, FdfHighlight, 'child') end

    local instance = child or Class.allocate(FdfHighlight)
    instance = FdfData.new(name, 'HIGHLIGHT', false, nil, instance)

    return instance
end

--========
-- Public
--========

---@param width number
function public:setWidth(width)
    isTypeErr(self, FdfHighlight, 'self')
    isTypeErr(width, 'number', 'width')

    local str_width = tostring(width)
    self:setParameter('Width', str_width)
end

---@param height number
function public:setHeight(height)
    isTypeErr(self, FdfHighlight, 'self')
    isTypeErr(height, 'number', 'height')

    local str_height = tostring(height)
    self:setParameter('Height', str_height)
end

--- Look-Up FileNames in some String table (for example gameinterface)
---@param flag boolean
function public:setDecorateFileNames(flag)
    isTypeErr(self, FdfHighlight, 'self')
    isTypeErr(flag, 'boolean', 'flag')

    if flag then
        self:setParameter('DecorateFileNames')
    else
        self:removeParameter('DecorateFileNames')
    end
end

---@param h_type string | "'FILETEXTURE'" | "'SHADE'"
function public:setHighlightType(h_type)
    isTypeErr(self, FdfHighlight, 'self')
    isTypeErr(h_type, 'string', 'h_type')

    self:setParameter('HighlightType', '\"'..h_type..'\"')
end

---@param path string
function public:setAlphaFile(path)
    isTypeErr(self, FdfHighlight, 'self')
    isTypeErr(path, 'string', 'path')

    self:setParameter('HighlightAlphaFile', '\"'..path..'\"')
end

---@param mode string | "'ADD'" | "'BLEND'"
function public:setAlphaMode(mode)
    isTypeErr(self, FdfHighlight, 'self')
    isTypeErr(mode, 'string', 'mode')

    self:setParameter('HighlightAlphaMode', '\"'..mode..'\"')
end

--- [0,1]
---@param r number
---@param g number
---@param b number
---@param a number
function public:setColor(r, g, b, a)
    isTypeErr(self, FdfHighlight, 'self')
    isTypeErr(r, 'number', 'r')
    isTypeErr(g, 'number', 'g')
    isTypeErr(b, 'number', 'b')
    isTypeErr(a, 'number', 'a')

    r = r < 0 and 0 or r > 1 and 1 or r
    g = g < 0 and 0 or g > 1 and 1 or g
    b = b < 0 and 0 or b > 1 and 1 or b
    a = a < 0 and 0 or a > 1 and 1 or a

    self:setParameter('HighlightColor', string.format('%f %f %f %f', r, g, b, a))
end

--=========
-- Private
--=========

return static

File: src/FrameExt/Fdf/Highlight.ts
import { Color } from "../../Utils";
import { Fdf } from "../Fdf";

type HighlightType = 'FILETEXTURE' | 'SHADE'
type HighlightMode = 'ADD' | 'BLEND'

export class FdfHighlight extends Fdf {
    constructor(name: string){
        super(name, 'HIGHLIGHT', false)
    }

    public get width(){return this._width}
    public set width(w: number){
        this._setParam('Width', w.toString())
        this._width = w
    }

    public get height(){return this._height}
    public set height(h: number){
        this._setParam('Height', h.toString())
        this._height = h
    }

    public get decorateFileNames(){return this._decorate}
    public set decorateFileNames(flag: boolean){
        if (flag){
            this._setParam('DecorateFileNames')
        } else {
            this._removeParam('DecorateFileNames')
        }
        this._decorate = flag
    }

    public get highlightType(){return this._highlight_type}
    public set highlightType(type: HighlightType){
        this._setParam('HighlightType', type)
        this._highlight_type = type
    }

    public get alphaFile(){return this._alpha_path}
    public set alphaFile(path: string){
        this._setParam('HighlightAlphaFile', path)
        this._alpha_path = path
    }

    public get alphaMode(){return this._alpha_mode}
    public set alphaMode(mode: HighlightMode){
        this._setParam('HighlightAlphaMode', mode)
        this._alpha_mode = mode
    }

    public get color(){return new Color(this._color)}
    public set color(c: Color){
        this._setParam('HighlightColor', string.format('%f %f %f %f', c.r, c.g, c.b, c.a))
        this._color = new Color(c)
    }

    private _width: number = -1;
    private _height: number = -1;
    private _decorate: boolean = false;
    private _highlight_type: HighlightType = 'SHADE';
    private _alpha_path: string = ''
    private _alpha_mode: HighlightMode = 'ADD'
    private _color: Color = new Color(1, 1, 1, 1)
}